### PR TITLE
[CPU] Add per-token asym INT8 dynamic quantization support to QKV/MLP node

### PR DESCRIPTION
**Pull request by usstq**
### Details:
 - to use AMX-INT8 to boost performance of QKV/MLP layers in LLM, we need dynamic per-token INT8 quantization according to many research papers (SmoothQuant for example) and reference implementations (xFT for example), here we add the support, when following conditions are met:
   - platform support AMX-INT8 
   - QKV&MLP weights are symmetrically per-OC quantized as INT8

- add AMX-FP16 support to QKV/MLP layers on GNR platform.
- optimize single batch 2nd token special case.

### Tickets:
 - *ticket-id*

### CODE REVIEW

**Comment by usstq** [comment]
Hi @dmitry-gorokhov , this is only a draft due to some unsolved design decision, like how to conditionally enable dynamic quantization, and maybe on per-layer basis. please take a look and comment, Thanks!

**Comment by usstq** [comment]
Hi @dmitry-gorokhov , this PR is ready for review, thanks!

**Comment by dmitry-gorokhov** [comment]
@luo-cheng2021 Could you please help to review the PR?

File: src/plugins/intel_cpu/src/nodes/qkv_proj.cpp [hunk 1]
```diff
@@ -277,22 +411,27 @@ bool QKVProjection::isSupportedOperation(const std::shared_ptr<const ov::Node>&
...
-            auto proj_pshape3 = op->input_value(3).get_shape();
-            if ((proj_pshape1[1] % CACHE_BLK_K_SIZE) != 0) {
+            const auto& config = node_qkv->get_config();
+            if ((config.hidden_size % CACHE_BLK_K_SIZE) != 0) {
                 errorMessage = "QKVProjection input channel size is not multiple of cache blocking size";
                 return false;
             }
-            if ((proj_pshape1[0] % REG_BLK_K_SIZE) != 0) {
+
+            if (config.quantized && (fcDynamicQuantizationGroupSize < static_cast<uint64_t>(config.hidden_size))) {
```
Comment (luo-cheng2021) [inline src/plugins/intel_cpu/src/nodes/qkv_proj.cpp:421]: Just for note: if the model is weight compressed but dynamic quantization not enabled, this check will force to the non-fused path.

File: src/plugins/intel_cpu/src/nodes/qkv_proj.cpp [hunk 2]
```diff
@@ -123,145 +157,245 @@ struct QKVProjection::Impl {
...
-        int strideA = srcStrides[1] * sizeof(ov::bfloat16);
-        auto stride_0 = dstStrides0[1];
-        auto stride_1 = dstStrides1[1];
-        auto stride_2 = dstStrides2[1];
+        int stride_src = srcStrides[1] * sizeof(T);
+        auto stride_dst_0 = dstStrides0[1];
+        auto stride_dst_1 = dstStrides1[1];
+        auto stride_dst_2 = dstStrides2[1];
 
+        auto asym = true;
```
Comment (luo-cheng2021) [inline src/plugins/intel_cpu/src/nodes/qkv_proj.cpp:222]: Just for note: here(+quantization in MatrixDynQuantPerRow) force the activation asym quantized.

File: src/plugins/intel_cpu/src/nodes/kernels/x64/mlp_kernel.cpp [hunk 3]
```diff
@@ -316,13 +493,49 @@ void MKernel::run(int M,  // actual M
...
         }
     }
 }
 
+void MatrixDynQuantPerRow::quantize(size_t BM, ov::bfloat16* psrc, int src_stride) {
+    assert(BM <= M);
+    parallel_nt_static(0, [&](const size_t ithr, const size_t nthr) {
+        size_t start{0}, end{0};
+        splitter(BM, nthr, ithr, start, end);
+        //auto prof = LinuxPerf::Profile("quant", start, end);
```
[the remark this answers is not in the document]
Comment (usstq) [reply src/plugins/intel_cpu/src/nodes/kernels/x64/mlp_kernel.cpp:508]: Sure

File: src/plugins/intel_cpu/src/nodes/kernels/x64/mlp_kernel.cpp [hunk 4]
```diff
@@ -316,13 +493,49 @@ void MKernel::run(int M,  // actual M
...
+                                                            asym);
+    });
+}
+
+void MatrixDynQuantPerRow::quantize(size_t BM, ov::float16* psrc, int src_stride) {
+    assert(BM <= M);
+    parallel_nt_static(0, [&](const size_t ithr, const size_t nthr) {
+        size_t start{0}, end{0};
+        splitter(BM, nthr, ithr, start, end);
+        //auto prof = LinuxPerf::Profile("quant", start, end);
```
[the remark this answers is not in the document]
Comment (usstq) [reply src/plugins/intel_cpu/src/nodes/kernels/x64/mlp_kernel.cpp:526]: Sure

File: src/plugins/intel_cpu/src/nodes/kernels/x64/mlp_kernel.hpp [hunk 5]
```diff
@@ -107,17 +151,56 @@ class MKernel : public dnnl::impl::cpu::x64::jit_generator {
...
+
+    using call_args = MKernel::call_args;
+
+    TMUL_TYPE m_tmul_type;
+    MKernel_1x2(TMUL_TYPE tmul_type) : jit_generator("MKernel_1x2") {
+        m_tmul_type = tmul_type;
+        create_kernel();
+    }
+
+    void tmul(const Xbyak::Tmm& x1, const Xbyak::Tmm& x2, const Xbyak::Tmm& x3) {
```
[the remark this answers is not in the document]
Comment (usstq) [reply src/plugins/intel_cpu/src/nodes/kernels/x64/mlp_kernel.hpp:174]: combined MKernel_1x2 into MKernel

File: src/plugins/intel_cpu/src/nodes/kernels/x64/mlp_utils.cpp [hunk 6]
```diff
@@ -17,12 +17,274 @@ namespace Cpu {
...
+#endif
+        for (; x < cols; x++) {
+            auto f_cur = static_cast<float>(psrc[x]);
+            f_min = std::min(f_min, f_cur);
+            f_max = std::max(f_max, f_cur);
+        }
+        // (q - z) * s = f
+        //  (-128 - z) * s = f_min;
+        //  ( 127 - z) * s = f_max;
+                float scale, zp;
```
[the remark this answers is not in the document]
Comment (usstq) [reply src/plugins/intel_cpu/src/nodes/kernels/x64/mlp_utils.cpp:70]: Sure

File: src/plugins/intel_cpu/src/nodes/kernels/x64/mlp_utils.cpp [hunk 7]
```diff
@@ -17,12 +17,274 @@ namespace Cpu {
...
+        p_scales[y] = scale;
+        x = 0;
+#if defined(HAVE_AVX512F)
+        auto vscale = _mm512_set1_ps(1.0f / scale);
+        auto vzp = _mm512_set1_ps(zp);
+        for (; x + 16 <= cols; x += 16) {
+            auto v = mm512_uni_loadu_ps(psrc + x);
+            v = _mm512_mul_ps(v, vscale);
+            v = _mm512_add_ps(v, vzp);
+            auto vi32x16 = _mm512_cvtps_epi32(v);
```
[the remark this answers is not in the document]
Comment (usstq) [reply src/plugins/intel_cpu/src/nodes/kernels/x64/mlp_utils.cpp:103]: Good idea, changed!

File: src/plugins/intel_cpu/src/nodes/kernels/x64/mlp_utils.cpp [hunk 8]
```diff
@@ -17,12 +17,274 @@ namespace Cpu {
...
+                              int dst_stride,
+                              int rows,
+                              int cols,
+                              float* p_scales,
+                              float* p_zp,
+                              bool asym) {
+    llm_mlp_quantize_to_i8(psrc, src_stride, pdst, dst_stride, rows, cols, p_scales, p_zp, asym);
+}
+
+/*
```
[the remark this answers is not in the document]
Comment (usstq) [reply src/plugins/intel_cpu/src/nodes/kernels/x64/mlp_utils.cpp:138]: Removed

File: src/plugins/intel_cpu/src/nodes/kernels/x64/mlp_utils.hpp [hunk 9]
```diff
@@ -2,17 +2,72 @@
...
+void llm_mlp_quantize_f16_i8(ov::float16* psrc,
+                              int src_stride,
+                              int8_t* pdst,
+                              int dst_stride,
+                              int rows,
+                              int cols,
+                              float* p_scales,
+                              float* p_zp,
+                              bool asym);
+/*
```
[the remark this answers is not in the document]
Comment (usstq) [reply src/plugins/intel_cpu/src/nodes/kernels/x64/mlp_utils.hpp:36]: Sure

File: src/plugins/intel_cpu/src/nodes/qkv_proj.cpp [hunk 10]
```diff
@@ -123,145 +157,245 @@ struct QKVProjection::Impl {
...
-            m_scratchMem = m_scrachPad->createScratchPadMem(newMemDesc);
+            if (m_node->m_config.quantized) {
+                m_quant_act.M = M;
+                m_quant_act.K = m_node->m_config.hidden_size;
+                allocator.register_allocation(m_quant_act.size(), [&](void* ptr){
+                    m_quant_act.setup(ptr);
+                });
+            }
 
+            // make sure scrach is big enough
```
[the remark this answers is not in the document]
Comment (usstq) [reply src/plugins/intel_cpu/src/nodes/qkv_proj.cpp:178]: Changed!

File: src/plugins/intel_cpu/src/nodes/qkv_proj.cpp [hunk 11]
```diff
@@ -123,145 +157,245 @@ struct QKVProjection::Impl {
...
+                            src,
+                            stride_src,
+                            m_quant_act.scale,
+                            m_quant_act.zp,
+                            work.w_sum_per_oc.template ptr<float>(),
+                            w_scale[work.output_id] + work.n0,
+                            asym);
+                    }
+                    // compress accumulation result into target
+                    //jit_cvt.call(src, stride_src, dst, stride_dst, BM, work.BN);
```
[the remark this answers is not in the document]
Comment (usstq) [reply src/plugins/intel_cpu/src/nodes/qkv_proj.cpp:281]: Sure

**Review by luo-cheng2021 — APPROVED** [review]
LGTM. Great job!

**Review by dmitry-gorokhov — APPROVED** [review]

**Comment by dmitry-gorokhov** [comment]
Just as general comment:
We will need to think about dynamic quantization flow generalization within the plugin. Curernt approach is to fully emdedd DQ logic inside compute operations (FC, MLP, QKV) which might good for performance (but not necessary required), but bad from scalability perspective.
Proper approach would be using separate DynamicQuantize operation (which to some extent is already presented in internal opset) and dedicated transformations which inserts corresponding DQ ops and transforms comptute ops in their quantized forms. 
That will allow to fully reuse quantization logic between the ops and make possible to fuse DQ ops as post ops in order to minimize memory traffic.

**Pull request merged**